datetime: fixed parse of time with h.hh or m.mm

tmr-g opened the pull request:
- `datetime.c`, `parse_tz_suffix` and its clients refactored
  for better error detection & to simplify code:
  - offset value post-condition check is moved to `parse_tz_suffix`.
  - this check uses tzoffset valid range instead of int16_t min/max.
  - offset arg type of `parse_tz_suffix` fixed
    to valid for TZoffset (int16_t).
- Fixed range checking for tzoffset parsed by `dt_parse_iso_zone_lenient`.
- The better error detection lead to assertion fails of datetime tests
  with decimal fraction of hours and minutes. The fix is the first
  way to solve the problem. (The other way is to comment out the tests.)

Requires tarantool/c-dt#3

Fixes #12082

---
@TarantoolBot document
Title: Parse of time with hour and minute decimal fraction implemented

Datetime module now supports formats, which are defined
in ISO 8601-1:2019 (5.3.1.4, b-c).

Implementation details:
  - As for decimal fraction of the second (5.3.1.4, a),
    the tail after 9 fraciton digits is truncated.
    `0.123456789999` is the same as `0.123456789`.
  - The hour (format c) or minute (format b) fractions are truncated
    to seconds precision. If somebody want a second fraction,
    they must use explicit representation (format a).

```
datetime.parse('2024-07-31T17,5', {format = 'iso8601'})
---
- 2024-07-31T17:30:00Z

datetime.parse('2024-07-31T17:30.333333333', {format = 'iso8601'})
---
- 2024-07-31T17:30:19Z
```

tmr-g committed:
datetime: fixed parse of time with h.hh or m.mm

- `datetime.c`, `parse_tz_suffix` and its clients refactored
  for better error detection & to simplify code:
  - offset value post-condition check is moved to `parse_tz_suffix`.
  - this check uses tzoffset valid range instead of int16_t min/max.
  - offset arg type of `parse_tz_suffix` fixed
    to valid for TZoffset (int16_t).
- Fixed range checking for tzoffset parsed by `dt_parse_iso_zone_lenient`.
- The better error detection lead to assertion fails of datetime tests
  with decimal fraction of hours and minutes. The fix is the first
  way to solve the problem. (The other way is to comment out the tests.)

Fixes #12082

@TarantoolBot document
Title: Parse of time with hour and minute decimal fraction implemented

Datetime module now supports formats, which are defined
in ISO 8601-1:2019 (5.3.1.4, b-c).

Implementation details:
  - As for decimal fraction of the second (5.3.1.4, a),
    the tail after 9 fraciton digits is truncated.
    `0.123456789999` is the same as `0.123456789`.
  - The hour (format c) or minute (format b) fractions are truncated
    to seconds precision. If somebody want a second fraction,
    they must use explicit representation (format a).

```
datetime.parse('2024-07-31T17,5', {format = 'iso8601'})
---
- 2024-07-31T17:30:00Z

datetime.parse('2024-07-31T17:30.333333333', {format = 'iso8601'})
---
- 2024-07-31T17:30:19Z
```

tmr-g committed:
datetime: fixed parse of time with h.hh or m.mm

- `datetime.c`, `parse_tz_suffix` and its clients refactored
  for better error detection & to simplify code:
  - offset value post-condition check is moved to `parse_tz_suffix`.
  - this check uses tzoffset valid range instead of int16_t min/max.
  - offset arg type of `parse_tz_suffix` fixed
    to valid for TZoffset (int16_t).
- Fixed range checking for tzoffset parsed by `dt_parse_iso_zone_lenient`.
- The better error detection lead to assertion fails of datetime tests
  with decimal fraction of hours and minutes. The fix is the first
  way to solve the problem. (The other way is to comment out the tests.)

Fixes #12082

@TarantoolBot document
Title: Parse of time with hour and minute decimal fraction implemented

Datetime module now supports formats, which are defined
in ISO 8601-1:2019 (5.3.1.4, b-c).

Implementation details:
  - As for decimal fraction of the second (5.3.1.4, a),
    the tail after 9 fraciton digits is truncated.
    `0.123456789999` is the same as `0.123456789`.
  - The hour (format c) or minute (format b) fractions are truncated
    to seconds precision. If somebody want a second fraction,
    they must use explicit representation (format a).

```
datetime.parse('2024-07-31T17,5', {format = 'iso8601'})
---
- 2024-07-31T17:30:00Z

datetime.parse('2024-07-31T17:30.333333333', {format = 'iso8601'})
---
- 2024-07-31T17:30:19Z
```

ligurio commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,66 @@ struct {
 		static char buff[DT_TO_STRING_BUFSIZE];
 		len = datetime_strftime(&date, buff, sizeof(buff), "%F %T%z");
-		ok(len > 0, "strftime");
+		ok(len > 0, "datetime_strftime");
```
why do you need this rename?
please revert. here and below

ligurio commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,66 @@ struct {
+	S("20121224T16:30.0+0100"),
+	SX("20121224T16:30.5+0100", DX(30)),
+	SX("20121224T16:30,1+0100", DX(6)),
+	SX("20121224T16:30,01666667+0100", DX(1)),
```
all these tests passed after reverting fix

ligurio commented on .gitmodules:
```diff
@@ -41,7 +41,8 @@
 	path = third_party/c-dt
-	url = https://github.com/tarantool/c-dt.git
+	url = https://github.com/tmr-g/c-dt
+	branch = i12082
```
Don't forget to revert before the merge.
I recommend moving all temporary changes to a separate commit with mark "[TMP]" and before the merge remove commit. This will guarantee that you will not forget to drop temporary changes.

tmr-g replied:
Thanks! I made tmp commit.

ligurio reviewed (commented):
Thanks for the patch! Please see comments inline.

tmr-g commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,66 @@ struct {
-		   "for '%s'", buff);
+		parse_len = datetime_parse_full(&date_parsed, buff, len);
+		is(parse_len > 0, true, "correct datetime_parse_full "
+		   "return value for '%s'", buff);
```
It's a typo fix. The `size_t` len can not receive `ssize_t` result of `datetime_parse_full`.

ligurio commented on changelogs/unreleased/gh-12082-datetime-parse-h-m-fractions.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/datetime
+
+* Fixed parse of time with hour's or minute's decimal fraction (gh-12082).
```
ISO 8601-1:2019 declare decimal fraction of hour, minute or second as an optional ("may", not "must"):

> A decimal fraction of hour, minute or second may be included in an expression in accordance with
ISO 8601-1:2019, 5.3.1.4.

ISO 8601-1:2019, 5.3.1.4 also declare it as optional:

> 5.3.1.4
> Representations with decimalfraction
> A decimal fraction of hour, minute or second may be included in a representation. If included, lower
> order time scale components(if any) shall be omitted and the decimal fractionshall be divided from the
> integer part by the decimal sign. If the magnitude of the numberis less than one, the decimal sign shall
> be preceded by two zeros.

It means that proposed changes are not a fix, but a feature.

tmr-g replied:
But we already have positive tests for fractions which means we support this feature: https://github.com/tarantool/tarantool/blob/dcdb3ee83b3d6324011e704b5a3f4ee3e19bbf47/test/app-luatest/datetime_test.lua#L133-L137

The feature is not work, so it looks as a bug.

tmr-g commented on test/unit/datetime.c:
```diff
@@ -15,15 +15,17 @@
+#define DEFAULT_SAMPLE_EPOCH 1356363000 /* 2012-12-24T15:30Z */
+#define S(s) {s, sizeof(s) - 1, DEFAULT_SAMPLE_EPOCH}
+#define SX(s, sample_epoch) {s, sizeof(s) - 1, sample_epoch}
+#define DX(delta) (DEFAULT_SAMPLE_EPOCH + (delta))
```
Your proposal is to rename `S` and all existing tests with `S`, which are not a part of this PR?
`SX` & `DX` will fix

tmr-g commented on test/unit/datetime.c:
```diff
@@ -15,15 +15,17 @@
 cord_on_yield(void) {}
 
-#define S(s) {s, sizeof(s) - 1}
+#define DEFAULT_SAMPLE_EPOCH 1356363000 /* 2012-12-24T15:30Z */
```
Done

tmr-g commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,68 @@ struct {
+
+		parse_len = datetime_parse_full(&date, test_str, test_len);
+		is(parse_len > 0, true, "correct datetime_parse_full "
+		   "return value for '%s'", test_str);
```
Thanks for reminder) I know that 'return value of `datetime_parse_full` is the number of parsed symbols'. That is your proposal?

You want to rename `parse_len`? You want to rewrite `is` text with "return value", inherited from the changed code?

tmr-g commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,68 @@ struct {
-		   "correct parse_datetime output "
+
+		parse_len = datetime_parse_full(&date, test_str, test_len);
+		is(parse_len > 0, true, "correct datetime_parse_full "
```
> So having positive parse_len is enough for success?

I suppose this inherit the same as `strptime` behavior, where the using of position of last parsed symbol depends on you goals. Returning 'parsed len' without success condition by `datetime_parse_full` is useless in this case. And it feels not nice.

Yes, here we must check equality of `parsed` with `test` as success condition, lets fix it.

tmr-g commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,68 @@ struct {
-		is(len > 0, true, "correct datetime_parse_full return value "
-		   "for '%s'", buff);
+		parse_len = datetime_parse_full(&date_parsed, buff, len);
+		is(parse_len > 0, true, "correct datetime_parse_full "
```
Fixed

tmr-g committed:
datetime: refactored to use MIN/MAX timezone offset constants

Constants `MIN_TZOFFSET` and `MAX_TZOFFSET` are introduced
in `datetime.lua` and in Lua test modules for better evolvability.
Also, they are used in C sources instead of literals.

NO_CHANGELOG=no visible changes
NO_TEST=no new behavior
NO_DOC=no new behavior

tmr-g commented on test/unit/datetime.c:
```diff
@@ -102,46 +104,68 @@ struct {
+	plan(476);
 
 	for (index = 0; index < lengthof(tests); index++) {
+		note("%s[%zu]", __func__, index);
```
Done

tmr-g commented on test/unit/datetime.c:
```diff
@@ -15,15 +15,17 @@
 #define UNIT_TAP_COMPATIBLE 1
```
Expected timestamps for fractions added.

tmr-g commented on changelogs/unreleased/gh-12082-datetime-parse-h-m-fractions.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/datetime
+
+* Fixed parse of time with hour's or minute's decimal fraction (gh-12082).
```
Done

ligurio commented on test/app-luatest/datetime_test.lua:
```diff
@@ -132,18 +132,23 @@ local SUPPORTED_DATETIME_FORMATS = {
         }, {
             fmt = '%Y-%M-%DT%,1h',
             buf = '2024-07-31T17,5',
+            ts = 1722447000,
```
let's define a variable with self-explained name for `1722447000,`

tmr-g commented on changelogs/unreleased/gh-12082-datetime-parse-h-m-fractions.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/datetime
```
Done

ligurio reviewed (approved):
Thanks for the patch! LGTM

tmr-g commented:
> Requires [gh-12417](https://github.com/tarantool/tarantool/issues/12417) to be fixed to pass fuzzing tests.

We decided to remove relation between #12083 and #12417: the weakened post condition check in `parse_tz_suffix` for timezone case is used.

ligurio commented on test/unit/datetime.c:
```diff
@@ -101,54 +129,106 @@ struct {
+{
+	const char *const invalid_tests[] = {
+		/* Out of range tzoffset. */
+		/* "8400329 10-22" case was found by fuzzer. */
```
I don't think it is meaningful information ("case was found by fuzzer"). It is not important how we found this bug/testcase.

tmr-g replied:
Removed.

Buristan commented on test/app-luatest/datetime_test.lua:
```diff
@@ -236,18 +262,23 @@ local SUPPORTED_DATETIME_FORMATS = {
         }, {
             fmt = '%Y-%M-%DT%,1hZ',
             buf = '2024-07-31T14,5Z',
+            ts = 1722436200,
```
IINM, there are two numeric constants here: `1722436200` and `1722447000`. I suggest naming them somehow to avoid magic numbers in the tests.

tmr-g replied:
Thanks! Done.

Buristan reviewed (approved):
Hi, Evgeniy!
Thanks for the patch!
LGTM, with minor nit below.
